finetune/util/group_metrics.py

```diff
@@ -0,0 +1,522 @@
+            costs[i][j] =  -(TP - (FP + FN) / 2)
+
+    pred_idxs, label_idxs = linear_sum_assignment(costs)
+    pred_idxs, label_idxs = list(pred_idxs), list(label_idxs)
+    assign_idxs = list(zip(pred_idxs, label_idxs))
+    return assign_idxs, counts
+
+def calc_class_counts(all_preds, all_labels, metric_type="group", span_type="exact"):
+    """
+    Calculate per-class TP, FP and FN counts.
+
+    :param all_preds, all_labels: A list of list of groups. If metric_type ==
+    "joint", the groups should in joint form with an "entities" key
+    :param metric_type: A string, either "group" or "joint"
+    :param span_type: A string, one of "exact", "token", "overlap", "superset"
+    :return class_counts: A dictionary of the form:
+        {
+            "class1": {
+                "true_positives: 1,
+                "false_positives: 1,
+                "false_negatives": 1,
+            }
+            .
+            .
+            .
+            "class_n: {...},
+        }
+    """
+    if metric_type == "joint":
+        assert "entities" in all_preds[0][0] and "entities" in all_labels[0][0], (
+            "To calculate joint metrics, groups must in in joint group format! "
+            "Use create_joint_groups() to produce groups in joint group format."
+        )
+
+    count_fn = get_count_fn(metric_type, span_type)
+    classes = _get_unique_classes(preds, labels)
+    class_counts = {
+        cls: {
+            "true_positives": 0,
+            "false_positives": 0,
+            "false_negatives": 0,
+        } for cls in classes
+    }
+
+    for doc_preds, doc_labels in zip(all_preds, all_labels):
+        if doc_preds and doc_labels:
+            assign_idxs, doc_counts = calc_group_assignment(
+                doc_preds, doc_labels, count_fn
+            )
+            for pred_idx, label_idx in assign_idxs:
+                TP, FP, FN = doc_counts[pred_idx][label_idx]
+                pred_label = doc_preds[pred_idx]["label"]
+                label_label = doc_labels[label_idx]["label"]
+
+                class_counts[label_label]["true_positives"] += TP
+                # False positives attributed to the predicted label
+                class_counts[pred_label]["false_positives"] += FP
+                # False negatives attributed to the ground truth label
+                class_counts[label_label]["false_negatives"] += FN
+
+        pred_idxs, label_idxs = list(zip(*assign_idxs))
+        if len(doc_preds) > len(doc_labels):
+            missing_idxs = set(range(len(doc_preds))) - set(pred_idxs)
+            extra_groups = doc_preds
+            # All examples in extra pred groups are false positives
+            add_count = "false_positives"
+        elif len(label_groups) > len(pred_groups):
+            missing_idxs = set(range(len(doc_labels))) - set(label_idxs)
+            extra_groups = doc_labels
+            # All examples in extra label groups are false negatives
+            add_count = "false_negatives"
```

Comment:
I don't think this accounts for the scenario where one class has a surplus of predictions and another has a deficit?

Reply:
Hmm, in that case I think it would be handled during the earlier group assignment step? We would do our best to match groups that have the same label but would probably be forced to make some assignments that aren't optimal. Regardless, I think it holds that all un-matched predictions are false positives and all un-matched ground truth labels are false negatives.

Reply:
Yup that checks out -- sorry for the false alarm, just took me a while to wrap my head around this one.